### PR TITLE
Big Tetris Parties

### PR DESCRIPTION
**Pull request by bitoffdev**
TL;DR This patch allows more than the previous 2 players to compete, and
makes many minor improvements.

Peek 2021-01-01 23-55

---

- Add magic number to all messages to help catch any network issues
- Add message_type field to message wire header, and remove the message_type
  from the body of many messages
- Define new method, message_type_to_str, to print a human-readable message
  type for logging
- Create new helper method message_blob
- Declare a macro name "PLAYER_NAME_MAX_CHARS", and use it consistently to
  validate the player names.
- Refactor StringArray to take a max length and define the StringArray
  serialized header bytes using a struct
- Bump MAXMSG up to 2048
- Refactor serialize_state to use no more bytes that necessary, rather than
  using MAXMSG as before
- Server now sends reply to user registration to indicate success
- Remove unnecessary and barely-utilized Player->io_lock, since all writes are
  done with a single call, which POSIX guarantees to be thread-safe.
- Add additional safety checks to renderer so that the program doesn't throw
  nasty errors (segfault) when the renderer has not been properly initialized
- Fix bug in read_from_client that referenced the buffer rather than the cursor

### CODE REVIEW

File: src/client_conn.c [hunk 1]
```diff
@@ -54,55 +56,50 @@ int get_socket() {
...
+	ttetris_net_request((NetClient *)net_client, NULL, 0,
+	                    MSG_TYPE_SWAP_HOLD);
 }
 
 StringArray *tetris_list(NetClient *net_client) {
-	char message[1];
-	message[0] = MSG_TYPE_LIST;
-	NetRequest *request = ttetris_net_request(net_client, message, 1);
+	NetRequest *request =
+	    ttetris_net_request(net_client, "", 0, MSG_TYPE_LIST);
```
Comment (Nitepone) [inline src/client_conn.c]: Curious. 2nd arg. NULL vs an empty string? Both seem to be used, is there a reason?

File: src/client_conn.c [hunk 2]
```diff
@@ -112,43 +109,47 @@ void tetris_connect(NetClient *net_client, char *host, int port) {
...
 
-int read_game_view_data(char **cursor, struct game_view_data *view) {
-	char *buffer = *cursor;
-	// move the pointer past the message identifier
-	++buffer;
+static int read_game_view_data(char *buffer, struct game_view_data *view) {
 	// get the player associated with the board
 	char *board_name = buffer;
-	int name_length = strlen(board_name);
+	unsigned long name_length = strnlen(board_name, PLAYER_NAME_MAX_CHARS);
```
Comment (Nitepone) [inline src/client_conn.c]: NIT: Agree with unsigned usage. But I hope that name length doesn't exceed the size of an int. Is there a reason to use a long here?

File: src/event.h [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/**
+ * simple library for describing events and waiting for events
```
Comment (Nitepone) [inline src/event.h:2]: This seems to currently be used primarily for game start and game stop. Is there plans to do more this this?

File: src/event.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#ifndef TTETRIS_EVENT_H
+#define TTETRIS_EVENT_H
+
+typedef struct ttetris_event TetrisEvent;
```
Comment (Nitepone) [inline src/event.h:8]: NIT: Having the struct and typedef separate is not something we have done much in our code. We have been defining structs like this entirely in the header files with a typedef. I believe that doing it this way will prevent usage of `struct ttetris_event` in files that import this header? Are there any other differences?

File: src/client.c [hunk 5]
```diff
@@ -97,12 +144,53 @@ int run_online(char *host, int port, char *username, char *opponent) {
...
+		selected_choice = selection_to_index(selection);
+
+		switch (selected_choice) {
+		case 0:
+			run_offline();
+			break;
+		case 1:
+			run_online(host, port);
+			break;
+		case 2:
```
Comment (Nitepone) [inline src/client.c:177]: Add a TODO here, And we should open up an issue for this.

File: src/client.c [hunk 6]
```diff
@@ -59,34 +65,75 @@ int run_list_online_players(char *host, int port) {
...
+	selection_destroy(selection);
+
+	return NULL;
+}
+
 /**
  * run an online game of tetris
  */
-int run_online(char *host, int port, char *username, char *opponent) {
+int run_online(char *host, int port) {
```
Comment (Nitepone) [inline src/client.c:97]: NIT: A TUI prompt for an address and port if `-a` is not specified should be added.

I'll create an issue for it if you agree.

File: src/client_conn.c [hunk 7]
```diff
@@ -54,55 +56,50 @@ int get_socket() {
...
+	NetRequest *request =
+	    ttetris_net_request(net_client, "", 0, MSG_TYPE_LIST);
 	ttetris_net_request_block_for_response(request);
 
 	MessageHeader *header = (MessageHeader *)request->cursor;
 
 	// deserialize names
 	Blob *body = malloc(sizeof(Blob));
-	body->bytes = request->cursor + 1;
+	body->bytes = request->cursor;
```
Comment (Nitepone) [inline src/client_conn.c:94]: Curious, what changed to not require this `+ 1`?

File: src/generic.c [hunk 8]
```diff
@@ -62,36 +71,45 @@ char *string_array_get_item(StringArray *arr, int index) {
...
 
 StringArray *string_array_deserialize(Blob *blob) {
-	int length = *((int *)blob->bytes);
-	StringArray *arr = string_array_create(length);
-
-	int message_index = 4;
-
-	for (int i = 0; i < length; i++) {
-		int string_length = strlen(blob->bytes + message_index);
+	// read the header first
```
Comment (Nitepone) [inline src/generic.c:101]: Some of this is getting a little complex. It might not be a bad idea to document how bits/bytes are used in these blobs?

**Review by Nitepone — COMMENTED** [review]
Some questions before approving this for merge. But, I don't see anything urgent to change.

File: src/client_conn.c (hunk 7 above)
Comment (bitoffdev) [reply src/client_conn.c:94]: Yeah; this is kind of confusing. It gets into the nitty-gritty of our message wire format.

The reason is because the first byte of the **message body** used to indicate the type of message (one of the `#declare`s in header.h such as `MSG_TYPE_REGISTER`). In this patch, I **moved the message type into MessageHeader**, which defines the first few bytes of every message.

Since the message type has been removed from the body, we no longer need to increment the cursor past it (The cursor is automatically pushed to the start of the body by logic in `read_from_client` and `read_from_server`).

In the future, I'd love to document our wire format a little so it's easy to understand when people find our repo. It's evolved a lot since the first version.

File: src/generic.c (hunk 8 above)
Comment (bitoffdev) [reply src/generic.c:101]: Yeah; that's fair. I'll create an issue for documenting the wire format.

The `StringArray` serializer might be the most complex that exists in our code right now, tbh. My hope is that introducing the [struct ttetris_string_array_wire_header](https://github.com/Nitepone/terminally-tetris/blob/88a277429027f9bbede8ecf12556cdd087616809/src/generic.c#L27) in this patch at least clears up the first few bytes. I've started using structs more to increase readability around our wire format when practical.

File: src/client.c (hunk 6 above)
Comment (bitoffdev) [reply src/client.c:97]: +1 Go ahead and create the issue; that makes sense to me! :smiley: 

In the future, I also envision that we might have a default server like ttetris.example.org if we get around to packaging, etc.

File: src/client.c (hunk 5 above)
Comment (bitoffdev) [reply src/client.c:177]: Good point. Implementing this can go under #21.

File: src/event.h (hunk 4 above)
Comment (bitoffdev) [reply src/event.h:8]: Preventing the usage of `struct ttetris_event` was my intent. My thought was to keep the definition opaque and force users to interact with `TetrisEvent`s through the interface defined by the methods in `header.h`. I believe this will make it easier to change the underlying functionality in the future if we want.

I'm not aware of any other differences.

File: src/event.h (hunk 3 above)
Comment (bitoffdev) [reply src/event.h:2]: You're correct that it's currently just for game start/stop, but it's super general. We could refactor some of the request-response networking code to use `event.h`. I am hoping it will be useful in the future as well, but I don't have any other specific ideas.

File: src/client_conn.c (hunk 2 above)
Comment (bitoffdev) [reply src/client_conn.c]: That's fair; I used a long because `strnlen` returns a `size_t`, which on most systems equates to an unsigned long afaik.

Of course, in practice `strnlen` will return a number smaller than `PLAYER_NAME_MAX_CHARS`, which is currently equal to 15, so I guess we could even use an `unsigned char`.

I will use `unsigned int` for now, per your suggestion. :smile:

File: src/client_conn.c (hunk 1 above)
Comment (bitoffdev) [reply src/client_conn.c]: Ahh; good catch. I'll update this to use NULL since an empty string is unnecessary bloat.

**Comment by bitoffdev** [comment]
Thanks for taking a look at my giant diff @Nitepone! :raised_hands: :heart_eyes: 

I think this latest commit should address your comments.

File: src/generic.c (hunk 8 above)
Comment (bitoffdev) [reply src/generic.c:101]: Created: #31

File: src/client_conn.c (hunk 7 above)
Comment (bitoffdev) [reply src/client_conn.c:94]: If you're curious, here is the [corresponding server change](https://github.com/Nitepone/terminally-tetris/pull/30/files/88a277429027f9bbede8ecf12556cdd087616809#diff-cbf91a87d3f07d2ad455ab70d64c16f55b7dc82be33346014bf25f91296d1e9eL40-L50) that removes the extra byte from the message body.

File: src/client_conn.c (hunk 2 above)
Comment (Nitepone) [reply src/client_conn.c]: That is interesting, didn't look at the output of strnlen!

File: src/event.h (hunk 3 above)
Comment (Nitepone) [reply src/event.h:2]: I see, sounds good to me! I bet we will find more uses for this.

File: src/event.h (hunk 4 above)
Comment (Nitepone) [reply src/event.h:8]: I fully agree with this rationale.

File: src/client.c (hunk 6 above)
Comment (Nitepone) [reply src/client.c:97]: See #33 

I do agree with this. Additionally, it may be cool to have a public server. However, we will need to take into account security risks before deploying such a thing.

File: src/client_conn.c (hunk 7 above)
Comment (Nitepone) [reply src/client_conn.c:94]: I see! Makes sense to me!

**Review by Nitepone — APPROVED** [review]
LGTM
